// @llamaindex/env/dist/cjs/index.workerd.js
/**
 * This module is under Cloudflare Workers environment.
 *
 * Most of Node.js APIs are not available in Cloudflare Workers environment.
 *
 * @module
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Tokenizers: function() {
        return _js.Tokenizers;
    },
    consoleLogger: function() {
        return _index1.consoleLogger;
    },
    emptyLogger: function() {
        return _index1.emptyLogger;
    },
    getEnv: function() {
        return getEnv;
    },
    loadTransformers: function() {
        return _indexnonnodejs.loadTransformers;
    },
    setTransformers: function() {
        return _indexnonnodejs.setTransformers;
    },
    tokenizers: function() {
        return _js.tokenizers;
    }
});
const _index = require("./utils/index.js");
_export_star(require("./node-polyfill.js"), exports);
const _index1 = require("./logger/index.js");
const _indexnonnodejs = require("./multi-model/index.non-nodejs.js");
const _js = require("./tokenizers/js.js");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
function getEnv(name) {
    return _index.INTERNAL_ENV[name];
}
