// @llamaindex/env/dist/cjs/index.browser.js
/**
 * Web environment polyfill.
 *
 * @module
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Tokenizers: function() {
        return _js.Tokenizers;
    },
    consoleLogger: function() {
        return _index.consoleLogger;
    },
    emptyLogger: function() {
        return _index.emptyLogger;
    },
    loadTransformers: function() {
        return _indexbrowser.loadTransformers;
    },
    setTransformers: function() {
        return _indexbrowser.setTransformers;
    },
    tokenizers: function() {
        return _js.tokenizers;
    }
});
require("./global-check.js");
_export_star(require("./web-polyfill.js"), exports);
const _index = require("./logger/index.js");
const _indexbrowser = require("./multi-model/index.browser.js");
const _js = require("./tokenizers/js.js");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
// @ts-expect-error
if (typeof window === "undefined") {
    console.warn("You are not in a browser environment. This module is not supposed to be used in a non-browser environment.");
}
